bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -509,7 +533,31 @@ func (r *KubeadmConfigReconciler) joinWorker(ctx context.Context, scope *Scope)
 			UseExperimentalRetry: scope.Config.Spec.UseExperimentalRetryJoin,
 		},
 		JoinConfiguration: joinData,
-	})
+	}
+
+	var cloudJoinData []byte
+
+	switch scope.Config.Spec.Format {
+	case bootstrapv1.Ignition:
+		ign, warnings, err := ignition.NewNode(&ignition.NodeInput{
+			NodeInput: nodeInput,
+			Ignition:  scope.Config.Spec.Ignition,
+		})
+
+		if err != nil {
+			scope.Error(err, "Failed to generate Ignition user data for the worker node")
+			return ctrl.Result{}, err
+		}
+
+		if warnings != "" {
+			conditions.MarkFalse(scope.Config, bootstrapv1.IgnitionGeneratedCondition, bootstrapv1.IgnitionHasWarningsReason, clusterv1.ConditionSeverityWarning, warnings)
```

Comment:
I'm a little bit confused by this condition.
It is kind of saying something went wrong, warning, but then the processing moves on no matter of the warning.
How the user is expected to notice this situation?
Also, even if it notices, what the user is expected to do given that the processing is already moved to the next stage?

Reply:
>Also, even if it notices, what the user is expected to do given that the processing is already moved to the next stage?

User can modify the configuration and roll an update to the machines. But they don't really have to, as those are only warnings. If strict mode is enabled, warnings are reported as errors, so provisioning won't proceed.

>How the user is expected to notice this situation?

IIUC conditions are emitted as events, so user can check there?

Reply:
> IIUC conditions are emitted as events

I'm not sure this is the case in CAPI.

However, the initial question holds, I'm still confused by the condition.
Please note that in CAPI all the conditions are expected to be summarized into the ready condition (ndr this is missing in this PR), and having a condition false will determine the KubeadmConfig to be ready=false, which will be weird given that the bootstrap secret is generated. 
Also KubeadmConfig being ready=false, will determine the Machine to be ready=false, thus creating more confusion.

In other words, according to our current rules conditions in CAPI are expected to report the operational state of a component, while here we are using a condition as a place for storing some warnings, which is an improper use of conditions.

As I was suggesting in other comments, IMO we should instead manage warning during CRD validation level.
If strict mode is enabled, in case of warning validation should fail and block object creation. If strict mode is disabled, we should ignore warnings and create the object.
As a result, at the time of creating the DataSecret object, there will be a guarantee that the data secret could be created because the user input is correct.

Reply:
>However, the initial question holds, I'm still confused by the condition.
Please note that in CAPI all the conditions are expected to be summarized into the ready condition (ndr this is missing in this PR), and having a condition false will determine the KubeadmConfig to be ready=false, which will be weird given that the bootstrap secret is generated.
Also KubeadmConfig being ready=false, will determine the Machine to be ready=false, thus creating more confusion.

Thanks for explaining this, I understand how this cause the confusion now. I wasn't aware that condition will affect overall readiness status. I thought this could be a way to inform user about the possible warnings in the configuration.

In order to be able to validate this data in the webhook, I must move my `ignition` package outside `internal` package, to be able to import it in the API. It will also make a import cycle, as `ignition` package already imports `bootstrapv1` package :thinking: 

>If strict mode is enabled, in case of warning validation should fail and block object creation. If strict mode is disabled, we should ignore warnings and create the object.
In one of other comments, someone suggested that there should be a way to inform user about the warnings. I couldn't see how this could be done on a webhook level, hence I tried with the conditions. Do you know if there is a way to produce a warning from the webhook (via controller-runtime, which is used I assume)?

Reply:
We should ask if controller-runtime supports warnings headers, or when this is planned

Reply:
> We should ask if controller-runtime supports warnings headers, or when this is planned

In the webhooks you can construct an AdmissionResponse with Warnings using a modifier func, this would be a good addition to notify users earlier that there's an issue.

 https://github.com/kubernetes-sigs/controller-runtime/blob/1d02366cf7fcca0c3993effbafea1d78cc48f55a/pkg/webhook/admission/response.go#L116-L121

Reply:
Hmm, it seems this is possible, but in order to make use of it, I must drop usage of `SetupWebhookWithManager` and create a handler myself. Do you think this is a good way to go? It it a requirement for this PR to get merged? Perhaps we could skip the warnings for now, or hardcode strict mode?

Reply:
If this is required, I think this work should be done in a separate PR before this one to limit the scope of changes here.

Reply:
I'm ok with moving the warning management to another PR

Reply:
Created https://github.com/kubernetes-sigs/cluster-api/issues/4279, so we can follow-up on warnings there.